### ISSUE TITLE
Always duplicate internal subresources on node duplicate (except when it does not make sense to).

### ISSUE THREAD

**Issue by reduz** [godotengine/godot-proposals#4672]
### Describe the project you are working on

Godot

### Describe the problem or limitation you are having in your project

It happens too often that Godot users are working on something that requires subresources (like particles, particle materials, materials, etc), then when duplicating the node the fact that these subresources are shared is entirely unexpected (you modify the subresource in one node and it changes in others).

Even if expected, it's very hard to track what is referenced internal within a scene. In contrast, users understand fine that if you reference a resource saved in a file (not internal), its always the same one even if you duplicate.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The idea is that, on the node duplication within the editor, if the resource is internal it should be almost always copied (non referenced) and if its external (saved to file) it will be referenced. This is far more intuitive for users and should avoid surprises. 

The problem with this is that there are some resource types that, even if internal, are quite expensive to have duplicated. They may consume a lot of resources for the user if they duplicate them without actually realizing that. Its the case of:

* Meshes
* Textures
* Materials

So, the proposed solution for this is to have a setting in the editor settings dialog that that lists the types that by default are not copied (and users have to use the Make Unique button).

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

* When duplicating, if the resource is not saved to disk, it will be copied.
* If added to the "forced share" list (that comes with some types by default), they are not copied, users have to use "Make Unique" if they _really_ want this.
* When duplicating resources within the above list, a warning popup or stack message that can be shown for the first time (that can be later dismissed) saying "Resources of type XXX will remain shared by default because this is more optimal for the engine. If you need them to be separate, use the 'Make Unique' option in the resource menu.".

### If this enhancement will not be used often, can it be worked around with a few lines of script?

N/A

### Is there a reason why this should be core and not an add-on in the asset library?

N/A

**Comment by h0lley** [comment]
I think it makes sense, but ultimately what would be the real game changer would be [a better visualization of sub-resources](https://github.com/godotengine/godot-proposals/issues/4567#issuecomment-1133602700) in the editor UI.

**Comment by reduz** [comment]
@h0lley I think a popup to see this could be useful (I originally tried this in the scene tree but its very messy so I removed it. What remains is RMBing), but the main problem to solve in this proposal is resources being shared without the user noticing.

**Comment by eon-s** [comment]
Well, this may be a source of misleading metrics when having performance issues for not reusing Resources (and not to mention how much scene files will grow this way), but if is optional and maybe with a shortcut shallow clone/duplicate at least, may be fine.

A better long term approach could be to force the creation of resources in files (scene subfolders or something like that) by default.

**Comment by reduz** [comment]
@eon-s Well, this is why we have an exclusion list of the resources that can really cause bloat if copied (and maybe a warning is shown). Encouraging the creation of resources in files is also a good idea I think, like we have for things like scripts and shaders. IMO it should be done for materials too.

**Comment by dploeger** [comment]
I absolutely agree with this as I've had multiple facepalm events because of this as well. 

I'd like to add that we should aim for consistency here. Maybe make **everything** duplicated per default and give the aforementioned options to speed up (with the given documentation)

It would be confusing for people if only _some_ resources duplicate by default.

**Comment by Riteo** [comment]
I kinda agree with @dploeger. What would the duplicable list be compared to the heavy stuff?

**Comment by reduz** [comment]
@Riteo @dploeger The list is in the proposal itself.

**Comment by RandomShaper** [comment]
I agree on the consistency goal. I think that after a node copy I'd have to check what has happened to the resources because the list may not be easy to interiorize. What if the list is merely for warning purposes? I mean, resources are copied by default and, when some of the types in the list is involved in a copy, a dialog tells you, 'The following resources have been copied, which are potentially heavy. Please consider sharing them to avoid bloat in scene files' and shows a list of the affected ones as node path + property + resource name/type. The dialog could also have a button to store them in files or just keep them as references to an unique instance.

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot-proposals/issues/4503.

**Comment by RandomShaper** [comment]
I didn't consider materials in my comment above. @Calinou is right. My idea still stands, but reversed: share by default, but allow in the dialog to make copies of the non-heavy resources (informed by the list).

**Comment by Riteo** [comment]
@reduz I mean, how big would the *duplicable* (eg. the lightweight stuff) list be? If it's big enough I guess that it may be easier to accept these cases as exceptions, although I'd still argue for consistency.

**Comment by QbieShay** [comment]
I think materials should be duplicated by default if they are spatial materials.

**Comment by tinmanjuggernaut** [comment]
I don't think it's a good idea. The engine should generally be more optimal by default, not wasteful by default. I don't want materials, textures, collision shapes, meshes, shaders to duplicate themselves in memory if I'm just making another instance of them. 

Unintentional duplicate materials would be the worst, as all the redundant textures will waste vram, and since the engine already has issues halting when compiling, more materials compounds the issue. 

What would be a good idea is changing the UI so shared resources have a small link icon on them. And clicking the icon `makes unique`, so I don't have to go into the context menu. 

Newbies are not the primary audience, as I've written before. Every new person eventually joins the middle of the userbase as they learn the quirks. So changing the workflow to accommodate newbies should never come at the cost of becoming less optimal by default.

I do agree that we need better resource visualization. The vram table is great. Let's make something like that for all resources. And one for all materials. And built in tools for the engine to precompile materials on demand. And on screen shader complexity visualization and other metrics (see UE for ideas). Overall we need better resource management, not more redundant resources which this change will create for everyone.

**Comment by noidexe** [comment]
I've been teaching Godot for a while and modifying a shared CollisionShape when you just wanted to duplicate it seems to be the Godot equivalent of drawing in the wrong layer in Photoshop 😄.  It can happen even to experienced devs from time to time. So I agree that having some resources duplicate by default is a good idea.  

On top of that, for the cases where I want something to be shared it'd be good to have an option to set it per resource rather than an editor-wide or project-wide setting per resource type. And in fact, there is a way to do it, but it's poorly communicated and doesn't work as expected. That is the `resource_local_to_scene` property.

If you don't know about the option, you probably won't ever stumble upon it since it's hidden inside the collapsed "Resource" section. If you do find it, the name is not very descriptive to what it does.

Even if you know about it, you may not know that instancing the same scene twice duplicates the resource, but instancing the scene once and then duplicating it with Control-D doesn't. Of course it makes sense from a programming standpoint, the property values are copied so all the references are copied, but it's really unexpected for the end user. 

I think it'd be much easier to understand for the end user if it were something like this:

![image](https://user-images.githubusercontent.com/526829/173930522-5f9d5f10-44d8-4b31-94be-a4bf6553fc00.png)

"Local To Scene" can be renamed as just "Shared" or  "Shared" and have a consistent reference vs dupicate behavior whether you duplicate a node, duplicate a scene instance or instance a scene. Is there a technical reason why it's hidden under Resource or why Resource is collapsed by default?

I can make it a separate proposal if you want but I think it's part of the same workflow

**Comment by bodamat** [comment]
Great idea, because it is really sometimes so annoying. What about adding an extra option for a duplicate with shared subresources? So, when you click the right mouse to the node you will have 2 options:
- Duplicate (Ctrl+D)
- Duplicate with shared subresources (Ctrl+Shift+D)

**Comment by noidexe** [comment]
> Great idea, because it is really sometimes so annoying. What about adding an extra option for a duplicate with shared subresources? So, when you click the right mouse to the node you will have 2 options:
Yeah I think Blender has something like that. If I'm not mistaken "Paste Objects" creates a copy of the materials while "Duplicate Objects" keeps the references to the same materials.

> I also completely agree with @noidexe. But could show (shared) text every time not when hovering.
I meant to say that (shared) always appears if the resource has more than one "owner" (Blender calls it "users"). It'd be extra nice if hovering on it showed you which nodes/scenes use that resource but maybe it's too much.

**Comment by h0lley** [comment]
> It'd be extra nice if hovering on it showed you which nodes/scenes use that resource but maybe it's too much.

I think the ability to understand at a glance what resources are used where is more important / helpful than the custom duplication behavior proposed. I'm still in favor of this proposal, just saying we additionally really should want better visualization, it's not too much. however, a tooltip would likely not suffice as it can't be made interactive. let's continue that in a separate proposal such as #4567 though.

**Comment by JyveAFK** [comment]
> > It'd be extra nice if hovering on it showed you which nodes/scenes use that resource but maybe it's too much.

To quickly see /something/, it should be the icon itself in the node tree denoting if it's a copy or a ref, with hovering over it giving more info of it's parent item/other nodes/scenes making use of that item, make it quicker to decide if you want to leave as is, or the right click from there makes it a unique copy.   

I'd want to look at a scene and quickly see which nodes are ref'd rather then click each node/hover over to figure that out at first.

Don't know if that'd be an easy thing to show... an underline of the icon? to a node to show it's a ref'd item.  

![ref](https://user-images.githubusercontent.com/32004318/173948253-111a6841-d599-43bc-b03b-94142fd79b93.png)

**Comment by Zylann** [comment]
When it comes to the *editor*, I would tend to agree on that internal resources should be duplicated most of the time. When the resource is in a file, it is very clear that it should be referenced, while an internal one is usually internal in the first place for being specific to an instance.

Problematic resources:
I'm not sure meshes or textures should be a problem in practice, we need to see what use cases these come up as internal.
Internal scripts are an odd one here once again tho: they are not just placed in Godot, they are also edited more deeply through the script editor, so them becoming duplicated can be very confusing and noticed late. Personally, I always hated them and avoided them, so I have no new solution to propose here...

> Unintentional duplicate materials would be the worst, as all the redundant textures will waste vram, and since the engine already has issues halting when compiling, more materials compounds the issue.

About nesting:
If the duplicated node has an external resource containing internal sub-resources, the resource should be referenced, and the sub-resources should NOT be duplicated. i.e recursion stops on the the first external.
If the duplicated node has an internal resource which contains external sub-resources, the resource should be duplicated, but its external sub-resources should stay referenced, again stopping recursion.
So in the case of materials, textures most of the time come from files, so there won't be duplication.

About importing:
Importing 3D scenes can come with resources being internal. 3D scenes, in their usable form (instances), are not resources. That makes duplicating instances also duplicate their resources right off the bat... so you'd really have to make sure importing uses files, and tbh, that's what I always go for regardless because it is clearer to navigate and makes it easier to re-use stuff like materials.

Note comparing to Unity, because I can't help it:
In Unity, all resources are files. There are no internal resources. Instead there are components and nestable object exports, which always copy naturally. Resources aren't used as widely as in Godot. Consequence is, this whole ambiguity doesn't exist here. And when a resource needs duplication, it will be a new file, which is clear and has some other benefits in asset navigation and referencing. In Godot, even a cube shape or rectangle can be a resource, and they can be internal, nested in others, even scripts are treated this way... which raises problems far more challenging to solve.

**Comment by wareya** [comment]
Would it be possible to somehow CoW internal resources?

**Comment by Zireael07** [comment]
@h0lley and the others discussing showing shared vs duplicated display: that belongs in https://github.com/godotengine/godot-proposals/issues/904, go in there and show that it's needed!

**Comment by graugraugrau** [comment]
I would agree. By copying or duplicating a node I would expect a deep copy.
Additional it would be nice if there would be an indicator if one resource is used by more than one scene. That would at least make it easier to detect. Maybe something like that:

![godot_proposal](https://user-images.githubusercontent.com/15648320/174067599-6258ed96-d6c6-49b5-95c7-ee1fcd627460.png)

**Comment by SlugFiller** [comment]
I've been personally burned on collision shapes. It makes sense for multiple runtime instances of an object (e.g. projectiles, enemies, etc) to share the collision shape instead of duplicating it, small as it may be. But if you're duplicating an object, that is not a scene, in the editor, you're generally expecting it to be a whole new object.

Meshes are not an example for an exception, either. Rather, one must ask, what is the use case where you duplicate an object with a mesh in the editor, and not wrap it in a scene? If it's a prefab of some sort, you would definitely put it in its own scene, so it can also have an associated collision shape and physics object. I can't imagine a case of using the same mesh twice in a scene.

As for materials, it certainly makes sense for textures to be shared, but very much NOT so for materials. If you have multiple objects with the same shader material, for example, you're likely interested in having different shader parameters applied. Things like tint for player color also comes to mind.

The worst case scenario is if an internal resource is copy and pasted across scenes. Godot does NOT handle this case gracefully. A resource that is internal to one scene, being referenced from another scene? It's an invitation for disaster. By all rights, this should be impossible to achieve, and be considered an invalid state. And yet, this is the default behavior for a simple copy and paste.

So I agree with what many others said above: If a resource is meant to be shared, it should be a file. Any other resource should be scene and object local. And should make it as simple as possible to determine and set whether it is instance local as well.

**Comment by TokisanGames** [comment]
Materials should be shared as much as possible as well. For unique variables there are uniform instances now in gd4. 
https://godotengine.org/article/godot-40-gets-global-and-instance-shader-uniforms

**Comment by Calinou** [comment]
I've started working on this: https://github.com/Calinou/godot/tree/editor-duplicate-resources-on-node-duplicate

https://github.com/godotengine/godot-proposals/assets/180032/e67b5f4b-7f5c-4941-b403-9fe4e8a0bc7d

Unfortunately, it doesn't always work reliably in all situations for reasons that elude me. Attempting to use it with nodes such as MeshInstance3D will print the following to console:

```
ERROR: Condition "!is_inside_tree()" is true. Returning: Transform3D()
   at: get_global_transform (./scene/3d/node_3d.cpp:343)
```

This seems to be simply caused by looping over the properties and checking whether they extend from Resource – maybe this is due to the `global_transform` property in Node3D? Either way, I'd need to exclude properties that cause known errors before attempting to detect which properties are resources.

**Comment by CookieBadger** [comment]
I philosophized a bit about the duplication dilemma here https://github.com/godotengine/godot-proposals/issues/317#issuecomment-3139824078, particularly about collision shapes, and noticed that their are two reasons two create resources:

- A: you want (potentially large amounts of) data be stored in a specific location for reuse (e.g. textures, materials, meshes)
- B: you want to be able to switch between different "modes of operation" of a node, each of which can expose a different set of parameters (e.g. the shape property of a CollsionShape3D node). 

In both cases, Resources are used, but in case B, users run into problems with duplicating, because the property values are expected to be specific to the instance of the node you are editing. Now, what about we separate the case B out from Resources into a new solution?

One quick idea would be a structure that we call "Mode" or "ParameterList" or whatever, which acts like a struct, i.e. it is always duplicated, and the data resides within each Node instance itself?

This way we wouldn't have to go the route of having an intransparent list of Resources that have custom behavior when duplicated...?

**Comment by idlist** [comment]
Just out of curiosity, would it make more sense to force shared subresources always be saved as a file? Personally I'm doing in this way for all my project so that there're zero in-node shared subresource, only saved ones.

**Comment by Calinou** [comment]
> Just out of curiosity, would it make more sense to force shared subresources always be saved as a file? Personally I'm doing in this way for all my project so that there're zero in-node shared subresource, only saved ones.

Some shared subresources are very small in file size and don't really benefit from being saved to a file, such as ButtonGroup or StyleBoxFlat. Otherwise, you'd have to save these to a file every time you create a radio button or a group of FoldableContainers.

**Comment by idlist** [comment]
>  Otherwise, you'd have to save these to a file every time you create a radio button or a group of FoldableContainers.

Though personally I do enforce the style of saving anything shared as a file (even in the case you mentioned) as I think the extra work is more worthwhile than having "ambiguously" shared resources, I do agree that if it's going to be a official thing, it would be a bit hard to design related process without being annoying in another direction.

Has been somebody troubled tracking shared and unique resources so I'm doing in this style. If it has the possibility to be an opt-in option then I'm totally going to tick it on, though again it's just sort of personal bias.

**Comment by h0lley** [comment]
> > Otherwise, you'd have to save these to a file every time you create a radio button or a group of FoldableContainers.
> 
> Though personally I do enforce the style of saving anything shared as a file (even in the case you mentioned) as I think the extra work is more worthwhile than having "ambiguously" shared resources, I do agree that if it's going to be a official thing, it would be a bit hard to design related process without being annoying in another direction.
> 
> Has been somebody troubled tracking shared and unique resources so I'm doing in this style. If it has the possibility to be an opt-in option then I'm totally going to tick it on, though again it's just sort of personal bias.

as with everything, there's a balance as just saving out all resources comes at the cost of cluttering your file system panel and adding a dependency to a tscn file that was potentially self-contained and possible to be copied across projects.

note that if you want to make things clearer, you can always name resources:

<img width="378" height="282" alt="Image" src="https://github.com/user-attachments/assets/8cc13c50-795d-495f-8b25-c227bc2b76b3" />

the name then appears in the inspector:

<img width="459" height="234" alt="Image" src="https://github.com/user-attachments/assets/d40b1ada-9aee-402d-bcc1-fc867ca26275" />

there's also this rather hidden sub resource explorer:

<img width="842" height="381" alt="Image" src="https://github.com/user-attachments/assets/38eedf07-a30c-4913-83b0-32795b485bce" />


lastly, when you name a built-in script resource, the name will also be used for the script list:

<img width="691" height="225" alt="Image" src="https://github.com/user-attachments/assets/22dcb274-1061-41c3-8166-7dd51653a4fc" />

**Comment by SlugFiller** [comment]
> > Just out of curiosity, would it make more sense to force shared subresources always be saved as a file? Personally I'm doing in this way for all my project so that there're zero in-node shared subresource, only saved ones.
> 
> Some shared subresources are very small in file size and don't really benefit from being saved to a file, such as ButtonGroup or StyleBoxFlat. Otherwise, you'd have to save these to a file every time you create a radio button or a group of FoldableContainers.

Something about this argument doesn't make sense to me. If they're not saved to a file, how are they shared? The OP is about the resource being auto-shared on duplication. Does that mean you need to use a very careful duplication chain to make sure these "small resources" are shared and not duplicated?

Also, this sharing might not work[^1], if the duplication is across scenes.

More importantly, could these "small resources" be too small to be a file, and yet, simultaneously, too large to just be duplicated instead of reused?

If I'm being honest, given how difficult[^2][^3] built-in resource management in Godot already is, I think "If it's built-in, it can't be shared" is a perfectly reasonable rule. Not only would it remove the foot-gun revolving duplication, it would also remove much of the general issues surrounding enumerating and inspecting built-in resources.

P.S. I used to dislike how the model exporter, by default, creates a file for each texture, instead of importing it as a built-in resource. Especially given how rare it is for two models to share a texture[^4]. But, in retrospect, making it clear that the potentially shared resource is reused and not duplicated, is ultimately good signaling.

[^1]: or even be extremely buggy
[^2]: and user unfriendly. e.g. when you want to know which resources are the same resource, and which are duplicates. Or when you want to know how many built-in resources a scene contains.
[^3]: and slightly buggy, especially if copy-pasting across scenes
[^4]: Most often, it would only be used once even within the model itself.
